source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,582 @@
+    {
+        return eAzureIoTHubClientTopicNoMatch;
+    }
+
+    /* Process incoming Publish. */
+    AZLogInfo( ( "Twin Incoming QoS : %d.", xMQTTPublishInfo->qos ) );
+
+    /* Verify the received publish is for the feature we have subscribed to. */
+    AZLogInfo( ( "Twin topic: %.*s. with payload : %.*s.",
+                 xMQTTPublishInfo->topicNameLength,
+                 xMQTTPublishInfo->pTopicName,
+                 xMQTTPublishInfo->payloadLength,
+                 xMQTTPublishInfo->pPayload ) );
+
+    if( pxContext->_internal.callbacks.xTwinCallback )
+    {
+        if( az_span_size( xOutRequest.xRequestID ) == 0 )
+        {
+            xTwinResponse.messageType = eAzureIoTHubTwinDesiredPropertyMessage;
+        }
+        else
+        {
+            if( az_result_failed( az_span_atou32( xOutRequest.xRequestID, &xRequestID ) ) )
+            {
+                return eAzureIoTHubClientFailed;
+            }
+
+            if( xRequestID & 0x01 )
+            {
+                xTwinResponse.messageType = eAzureIoTHubTwinReportedResponseMessage;
+            }
+            else
+            {
+                xTwinResponse.messageType = eAzureIoTHubTwinGetMessage;
+            }
+        }
+
+        xTwinResponse.pvMessagePayload = xMQTTPublishInfo->pPayload;
+        xTwinResponse.ulPayloadLength = ( uint32_t ) xMQTTPublishInfo->payloadLength;
+        xTwinResponse.messageStatus = xOutRequest.status;
+        xTwinResponse.pucVersion = az_span_ptr( xOutRequest.version );
+        xTwinResponse.usVersionLength = ( uint16_t ) az_span_size( xOutRequest.version );
+        xTwinResponse.ulRequestID = xRequestID;
+        pxContext->_internal.callbacks.xTwinCallback( &xTwinResponse,
+                                                      pxContext->_internal.pvCallbackContext );
+    }
+
+    return eAzureIoTHubClientSuccess;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Time callback for MQTT initialization.
+ *
+ * */
+static uint32_t prvGetTimeMs( void )
+{
+    TickType_t xTickCount = 0;
+    uint32_t ulTimeMs = 0UL;
+
+    /* Get the current tick count. */
+    xTickCount = xTaskGetTickCount();
+
+    /* Convert the ticks to milliseconds. */
+    ulTimeMs = ( uint32_t ) xTickCount * azureiotMILLISECONDS_PER_TICK;
+
+    return ulTimeMs;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_OptionsInit( AzureIoTHubClientOptions_t * pxHubClientOptions )
+{
+    AzureIoTHubClientResult_t xResult;
+
+    if( pxHubClientOptions == NULL )
+    {
+        AZLogError( ( "Failed to initialize options: Invalid argument." ) );
+        xResult = eAzureIoTHubClientInvalidArgument;
+    }
+    else
+    {
+        pxHubClientOptions->pucModelID = NULL;
+        pxHubClientOptions->ulModelIDLength = 0;
+        pxHubClientOptions->pucModuleID = NULL;
+        pxHubClientOptions->ulModuleIDLength = 0;
```

Comment:
memset?

Reply:
should we do it like the embedded?

https://github.com/Azure/azure-sdk-for-c/blob/fed9dd0b2a000ab7e607bac76112a983f2bd8829/sdk/src/azure/iot/az_iot_hub_client.c#L27-L32

```c
AZ_NODISCARD az_iot_hub_client_options az_iot_hub_client_options_default()
{
  return (az_iot_hub_client_options){ .module_id = AZ_SPAN_EMPTY,
                                      .user_agent = client_sdk_version,
                                      .model_id = AZ_SPAN_EMPTY };
}
```

compiler then gives you warning about uninitialized members

Reply:
I prefer not to pass structs by return values.